File: web/src/dicionarioLocal.ts
import Dexie, { type EntityTable } from 'dexie';
import {
  candidatos,
  colapsarRepeticoes,
  normalizar,
  tokenizar,
  type GiriaDetectada,
  type RespostaDeTraducao,
} from '@tradugiria/core';

/**
 * Nível 0 da cascata: o dicionário que vive no dispositivo.
 *
 * É o que faz o RNF07 valer — a consulta funciona sem conexão, e a Marlene
 * não gasta o plano de dados para entender uma mensagem. Também é o caminho
 * mais rápido: resolver aqui evita a ida à rede por completo.
 *
 * A detecção reusa `tokenizar` e `candidatos` do pacote compartilhado, de
 * modo que o texto é quebrado exatamente como o servidor quebraria. Se esta
 * função divergisse da do servidor, o mesmo texto produziria destaques em
 * posições diferentes conforme houvesse ou não internet.
 */

interface VerbeteLocal {
  termoNormalizado: string;
  termo: string;
  idioma: string;
  explicacaoSimples: string;
  explicacaoDetalhada: string | null;
  equivalenteFormal: string | null;
  nsfw: boolean;
  riscoMenor: boolean;
  /** Variações normalizadas que apontam para este mesmo verbete. */
  variacoes: string[];
}

const bancoLocal = new Dexie('tradugiria') as Dexie & {
  verbetes: EntityTable<VerbeteLocal, 'termoNormalizado'>;
};

bancoLocal.version(1).stores({
  // O índice multiEntry em `variacoes` permite achar o verbete por qualquer
  // apelido sem varrer a tabela inteira.
  verbetes: 'termoNormalizado, idioma, *variacoes',
});

export async function guardarVerbetes(verbetes: VerbeteLocal[]): Promise<void> {
  await bancoLocal.verbetes.bulkPut(verbetes);
}

export async function quantidadeGuardada(): Promise<number> {
  return bancoLocal.verbetes.count();
}

async function procurar(chave: string): Promise<VerbeteLocal | undefined> {
  const direto = await bancoLocal.verbetes.get(chave);
  if (direto) return direto;

  const porVariacao = await bancoLocal.verbetes.where('variacoes').equals(chave).first();
  if (porVariacao) return porVariacao;

  const colapsado = colapsarRepeticoes(chave);
  if (colapsado !== chave) {
    return bancoLocal.verbetes.get(colapsado);
  }
  return undefined;
}

/**
 * Traduz usando apenas o que está no dispositivo.
 *
 * Reproduz o casamento guloso do servidor: candidatos mais longos primeiro,
 * e nenhum caractere reclamado duas vezes — sem isso, "dar ruim" viria
 * acompanhado de "ruim" sozinho e a tela desenharia destaques sobrepostos.
 */
export async function traduzirLocalmente(
  texto: string,
  opcoes: { nivel: 'SIMPLES' | 'DETALHADA'; modoFamilia: boolean },
): Promise<RespostaDeTraducao> {
  const lista = candidatos(tokenizar(texto));
  const detectadas: GiriaDetectada[] = [];
  const naoResolvidos = new Set<string>();
  const ocupado = new Array<boolean>(texto.length).fill(false);

  for (const candidato of lista) {
    if (intervaloOcupado(ocupado, candidato.inicio, candidato.fim)) continue;

    const verbete = await procurar(candidato.normalizado);
    if (!verbete) {
      if (!candidato.normalizado.includes(' ') && candidato.normalizado.length >= 3) {
        naoResolvidos.add(candidato.normalizado);
      }
      continue;
    }

    if (opcoes.modoFamilia && verbete.nsfw) {
      marcarOcupado(ocupado, candidato.inicio, candidato.fim);
      continue;
    }

    detectadas.push({
      termo: verbete.termo,
      posicao: [candidato.inicio, candidato.fim],
      explicacao:
        opcoes.nivel === 'DETALHADA' && verbete.explicacaoDetalhada
          ? verbete.explicacaoDetalhada
          : verbete.explicacaoSimples,
      equivalenteFormal: verbete.equivalenteFormal,
      nsfw: verbete.nsfw,
      riscoMenor: verbete.riscoMenor,
      confianca: 1,
      origem: 'DICIONARIO',
    });
    marcarOcupado(ocupado, candidato.inicio, candidato.fim);
  }

  detectadas.sort((a, b) => a.posicao[0] - b.posicao[0]);

  return {
    idiomaDetectado: null,
    girias: detectadas,
    geradoPorIa: false,
    naoResolvidos: [...naoResolvidos],
  };
}

/**
 * Guarda no dispositivo o que o servidor acabou de resolver.
 *
 * É o que faz o dicionário local crescer com o uso: os termos que a pessoa
 * realmente consulta ficam disponíveis offline na próxima vez, sem precisar
 * baixar o pacote inteiro.
 */
export async function memorizar(resposta: RespostaDeTraducao): Promise<void> {
  const novos: VerbeteLocal[] = resposta.girias
    .filter((g) => g.origem === 'DICIONARIO' && g.explicacao !== null)
    .map((g) => ({
      termoNormalizado: normalizar(g.termo),
      termo: g.termo,
      idioma: resposta.idiomaDetectado ?? 'pt-BR',
      explicacaoSimples: g.explicacao!,
      explicacaoDetalhada: null,
      equivalenteFormal: g.equivalenteFormal,
      nsfw: g.nsfw,
      riscoMenor: g.riscoMenor,
      variacoes: [],
    }));

  if (novos.length > 0) {
    await guardarVerbetes(novos);
  }
}

function intervaloOcupado(ocupado: boolean[], inicio: number, fim: number): boolean {
  for (let i = inicio; i < fim; i++) {
    if (ocupado[i]) return true;
  }
  return false;
}

function marcarOcupado(ocupado: boolean[], inicio: number, fim: number): void {
  for (let i = inicio; i < fim; i++) {
    ocupado[i] = true;
  }
}

File: android/app/src/main/java/br/com/tradugil/android/dados/DicionarioLocal.kt
package br.com.tradugil.android.dados

import br.com.tradugil.android.dados.local.DicionarioDao
import br.com.tradugil.android.dados.local.VariacaoLocal
import br.com.tradugil.android.dados.local.VerbeteLocal
import br.com.tradugil.android.dominio.GiriaDetectada
import br.com.tradugil.android.dominio.NivelDeExplicacao
import br.com.tradugil.android.dominio.Normalizador
import br.com.tradugil.android.dominio.OrigemDaResposta
import br.com.tradugil.android.dominio.RespostaDeTraducao
import br.com.tradugil.android.dominio.Tokenizador

/**
 * Traduz usando apenas o que está no aparelho.
 *
 * <p>Reproduz o casamento guloso do servidor: candidatos mais longos
 * primeiro, e nenhum caractere reclamado duas vezes. Sem isso, "dar ruim"
 * viria acompanhado de "ruim" sozinho e a tela desenharia dois destaques
 * sobrepostos.</p>
 *
 * <p>Usa `Tokenizador` e `Normalizador` do domínio: as mesmas funções que o
 * servidor usa em Java e o site em TypeScript. É o que garante que o texto
 * seja quebrado nas mesmas posições com ou sem internet: se esta classe
 * tivesse a própria lógica, o mesmo texto produziria destaques em lugares
 * diferentes conforme a conexão.</p>
 */
class DicionarioLocal(private val dao: DicionarioDao) {

    private companion object {
        /** Termo curto demais para valer uma linha na fila de curadoria. */
        const val TAMANHO_MINIMO_PARA_FILA = 3
    }

    suspend fun quantidadeGuardada(): Int = dao.quantidade()

    suspend fun traduzir(
        texto: String,
        nivel: NivelDeExplicacao,
        modoFamilia: Boolean,
    ): RespostaDeTraducao {
        val candidatos = Tokenizador.candidatos(Tokenizador.tokenizar(texto))
        if (candidatos.isEmpty()) {
            return RespostaDeTraducao(null, emptyList(), false, emptyList())
        }

        val chaves = candidatos.map { it.normalizado }.distinct()
        val colapsadas = chaves.map { Normalizador.colapsarRepeticoes(it) }.distinct()

        // Indexado pelas quatro formas de alcance, para o laço de detecção
        // encontrar o verbete perguntando pela forma que ele tem em mãos.
        val porChave = mutableMapOf<String, VerbeteLocal>()
        dao.buscarPorCandidatos(chaves, colapsadas).forEach { verbete ->
            porChave[verbete.termoNormalizado] = verbete
            porChave.putIfAbsent(verbete.termoColapsado, verbete)
        }
        candidatos.forEach { candidato ->
            if (!porChave.containsKey(candidato.normalizado)) {
                porChave[Normalizador.colapsarRepeticoes(candidato.normalizado)]?.let {
                    porChave[candidato.normalizado] = it
                }
            }
        }

        val detectadas = mutableListOf<GiriaDetectada>()
        val naoResolvidos = linkedSetOf<String>()
        val ocupado = BooleanArray(texto.length)

        for (candidato in candidatos) {
            if (intervaloOcupado(ocupado, candidato.inicio, candidato.fim)) continue

            val verbete = porChave[candidato.normalizado]
            if (verbete == null) {
                if (!candidato.normalizado.contains(" ") &&
                    candidato.normalizado.length >= TAMANHO_MINIMO_PARA_FILA
                ) {
                    naoResolvidos += candidato.normalizado
                }
                continue
            }

            // Modo Família: o impróprio some, o de risco fica e vem marcado.
            if (modoFamilia && verbete.nsfw) {
                marcarOcupado(ocupado, candidato.inicio, candidato.fim)
                continue
            }

            detectadas += GiriaDetectada(
                termo = verbete.termo,
                inicio = candidato.inicio,
                fim = candidato.fim,
                explicacao = if (nivel == NivelDeExplicacao.DETALHADA &&
                    verbete.explicacaoDetalhada != null
                ) {
                    verbete.explicacaoDetalhada
                } else {
                    verbete.explicacaoSimples
                },
                equivalenteFormal = verbete.equivalenteFormal,
                nsfw = verbete.nsfw,
                riscoMenor = verbete.riscoMenor,
                confianca = 1.0,
                origem = OrigemDaResposta.DICIONARIO,
            )
            marcarOcupado(ocupado, candidato.inicio, candidato.fim)
        }

        return RespostaDeTraducao(
            idiomaDetectado = null,
            girias = detectadas.sortedBy { it.inicio },
            geradoPorIa = false,
            naoResolvidos = naoResolvidos.toList(),
        )
    }

    /**
     * Guarda o que o servidor acabou de resolver.
     *
     * <p>É o que faz o dicionário do aparelho crescer com o uso: os termos
     * que a pessoa realmente consulta ficam disponíveis offline na próxima
     * vez, sem precisar baixar o pacote inteiro.</p>
     *
     * <p>Só guarda o que veio do dicionário curado. Resposta de IA não entra:
     * ela é explicitamente não verificada, e gravá-la aqui a transformaria,
     * na consulta seguinte, num verbete offline indistinguível dos revisados.</p>
     */
    suspend fun memorizar(resposta: RespostaDeTraducao) {
        val verbetes = resposta.girias
            .filter { it.origem == OrigemDaResposta.DICIONARIO && it.explicacao != null }
            .map { giria ->
                val normalizado = Normalizador.normalizar(giria.termo)
                VerbeteLocal(
                    termoNormalizado = normalizado,
                    termoColapsado = Normalizador.colapsarRepeticoes(normalizado),
                    termo = giria.termo,
                    idioma = resposta.idiomaDetectado ?: "pt-BR",
                    explicacaoSimples = giria.explicacao!!,
                    explicacaoDetalhada = null,
                    equivalenteFormal = giria.equivalenteFormal,
                    nsfw = giria.nsfw,
                    riscoMenor = giria.riscoMenor,
                )
            }

        if (verbetes.isNotEmpty()) {
            dao.guardar(verbetes, emptyList<VariacaoLocal>())
        }
    }

    private fun intervaloOcupado(ocupado: BooleanArray, inicio: Int, fim: Int): Boolean {
        for (i in inicio until fim) if (ocupado[i]) return true
        return false
    }

    private fun marcarOcupado(ocupado: BooleanArray, inicio: Int, fim: Int) {
        for (i in inicio until fim) ocupado[i] = true
    }
}
